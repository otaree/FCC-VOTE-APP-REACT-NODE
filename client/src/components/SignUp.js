import React from 'react';
import { connect } from 'react-redux';
import isEmail from 'validator/lib/isEmail';
import axios from 'axios';


import * as authActions from '../store/actions/auth';

export class SignUp extends React.Component {
    state = {
        username: {
            value: "",
            error: false
        },
        email: {
            value: "",
            error: false
        },
        password: {
            value: "",
            error: false
        },
        password2: {
            value: "",
            error: false
        }
    };

    nameChangeHandler = e => {
        this.updateFieldValue("username", e.target.value);
    };

    emailChangeHandler = e => {
        this.updateFieldValue("email", e.target.value);
    };

    passwordChangeHandler = e => {
        this.updateFieldValue("password", e.target.value);
    };

    password2ChangeHandler = e => {
        this.updateFieldValue("password2", e.target.value);
    };

    submitHandler = async e => {
        e.preventDefault();

        if (this.validator()) return;

        const user = { 
            name: this.state.username.value, 
            email: this.state.email.value, 
            password: this.state.password.value  
        };

        try {
            const response = await axios.post("http://localhost:5000/user", user);
            this.props.signUp({ ...response.data, token: response.headers['x-auth'] });
            this.props.history.push('/');
        } catch (e) {
            this.props.signUpFail();
        }

    }

    validator = () => {
        let error = false;

        if (this.state.username.value < 1) {
            error = true;

            this.setState(prevState => {
                return {
                    username: {
                        value: prevState.username.value,
                        error: true
                    }
                };
            });
        }

        if (!isEmail(this.state.email.value)) {
            error = true;
            
            this.setState(prevState => {
                return {
                    email: {
                        value: prevState.email.value,
                        error: true
                    }
                };
            });
        }

        if (this.state.password.value.length < 6) {
            error = true;
            
            this.setState(prevState => {
                return {
                    password: {
                        value: prevState.password.value,
                        error: true
                    }
                };
            });
        }

        if (this.state.password.value !== this.state.password2.value) {
            error = true;
            
            this.setState(prevState => {
                return {
                    password2: {
                        value: prevState.password2.value,
                        error: true
                    }
                };
            });
        }

       return error;
    };

    updateFieldValue = (field, value) => {
        this.setState(prevState => {
            return {
                [field]: {
                    value,
                    error: false
                }
            };
        });
    };

    render() {
        return (
            <section className="auth">
                {
                    this.props.authError.isError &&
                    <div className="auth-server-fail">{this.props.authError.value}</div>
                }
                <form className="form" onSubmit={this.submitHandler}>
                    <div className="form-control">
                        <input 
                            type="text"
                            placeholder="Username"
                            value={this.state.username.value}
                            onChange={this.nameChangeHandler}
                        />
                        {
                            this.state.username.error &&
                            <div className="auth-error">Please provide a username</div>
                        }
                    </div>
                    <div className="form-control">
                        <input 
                             type="email"
                             placeholder="Email"
                             value={this.state.email.value}
                             onChange={this.emailChangeHandler}
                         />
                         {
                             this.state.email.error &&
                             <div className="auth-error">Invalid email</div>
                         }
                   </div>
                   <div className="form-control">
                        <input 
                             type="password"
                             placeholder="Password"
                             value={this.state.password.value}
                             onChange={this.passwordChangeHandler}
                         />
                         {
                             this.state.password.error &&
                             <div className="auth-error">Invalid password</div>
                         }
                   </div>
                   <div className="form-control">
                        <input 
                             type="password"
                             placeholder="Confirm Password"
                             value={this.state.password2.value}
                             onChange={this.password2ChangeHandler}
                         />
                         {
                             this.state.password2.error &&
                             <div className="auth-error">Passwords dose not match</div>
                         }
                   </div>
                   <div className="form-control">
                        <input 
                            className="btn"
                             type="submit"
                             value="Register"
                         />
                   </div>
                </form>
            </section>
        );
    }
}

const mapStateToProps = state => {
    return {
        authError: state.auth.error,
        loading: state.auth.loading
    };
};

const mapDispatchProps = dispatch => {
    return {
        signUp: (user) => dispatch(authActions.login(user)),
        signUpFail: () => dispatch(authActions.fail("Auth Error"))
    };
};

export default connect(mapStateToProps, mapDispatchProps)(SignUp);