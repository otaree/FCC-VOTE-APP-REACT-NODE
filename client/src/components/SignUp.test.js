import React from 'react';
import { shallow, mount } from 'enzyme';
// import axios from 'axios';
// import MockAdapter from 'axios-mock-adapter';

import { SignUp } from './SignUp';

// jest.mock('axios', () => {
//     const exampleArticles = [
//         { title: 'test article', url: 'test url' }
//       ];

//     return {
//         post:  jest.fn(() => Promise.resolve(exampleArticles)),
//     }
// });

// const axios = require('axios');

describe('SignUp', () => {
    let mock;
    const mockSignUp = jest.fn();
    const mockSignUpFail = jest.fn();
    const mockPush = jest.fn();

    let props = {
        signUp: mockSignUp,
        signUpFail: mockSignUpFail,
        history: {
            push: mockPush
        },
        authError: {
            isError: false,
            value: ''
        }
    }
    let signUp = shallow(<SignUp {...props} />);

    it('should renders correctly', () => {
        expect(signUp.find('.auth-server-fail').length).toEqual(0);        
        expect(signUp).toMatchSnapshot();
    });

    describe('When from fields value changes', () => {
        beforeEach(() => {
            signUp = shallow(<SignUp {...props} />);
        });

        it('should set username on input change', () => {
            const value = "Test";
            signUp.find('input').at(0).simulate('change', {
                target: { value }
            });
            expect(signUp.state('username').value).toEqual(value);
        });

        it('should set email on input change', () => {
            const value = "test@test.com";
            signUp.find('input').at(1).simulate('change', {
                target: { value }
            });
            expect(signUp.state('email').value).toEqual(value);
        });

        it('should set password on input change', () => {
            const value = "password!";
            signUp.find('input').at(2).simulate('change', {
                target: { value }
            });
            expect(signUp.state('password').value).toEqual(value);
        });

        it('should set password2 on input change', () => {
            const value = "password!";
            signUp.find('input').at(3).simulate('change', {
                target: { value }
            });
            expect(signUp.state('password2').value).toEqual(value);
        });
    });

    describe('when authentication Error', () => {
        beforeEach(() => {
            props.authError.isError = true;
            props.authError.value = "Authentication Error";
            signUp = shallow(<SignUp {...props} />);
        });

        it('should render the error message', () => {
            expect(signUp.find('.auth-error').length).toEqual(1);
        });
    });

    // describe('when valid form submission', () => {
    //     beforeEach(() => {
    //         signUp = shallow(<SignUp {...props} />);
    //         mock = new MockAdapter(axios);
    //     });

    //     it('should submit the from correctly', () => {
    //         const user = {
    //             name: "Test",
    //             email: "test@test.com",
    //             password: "password"
    //         };  

    //         signUp.setState({ username: { value: user.name, error: false }});
    //         signUp.setState({ email: { value: user.email, error: false }});
    //         signUp.setState({ password: { value: user.password, error: false }});
    //         signUp.setState({ password: { value: user.password2, error: false }});
    //         signUp.find('form').simulate('submit', {
    //             preventDefault: () => {}
    //         });

    //         expect(mockSignUpFail).toHaveBeenCalled();
                      
    //     });
    // });
});