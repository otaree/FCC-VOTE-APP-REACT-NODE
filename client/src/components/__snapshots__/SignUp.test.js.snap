// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SignUp should renders correctly 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <SignUp
    authError={
        Object {
            "isError": false,
            "value": "",
          }
    }
    history={
        Object {
            "push": [Function],
          }
    }
    signUp={[Function]}
    signUpFail={[Function]}
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        false,
        <form
          className="form"
          onSubmit={[Function]}
>
          <div>
                    <input
                              onChange={[Function]}
                              placeholder="Username"
                              type="text"
                              value=""
                    />
          </div>
          <div>
                    <input
                              onChange={[Function]}
                              placeholder="Email"
                              type="email"
                              value=""
                    />
          </div>
          <div>
                    <input
                              onChange={[Function]}
                              placeholder="Password"
                              type="password"
                              value=""
                    />
          </div>
          <div>
                    <input
                              onChange={[Function]}
                              placeholder="Confirm Password"
                              type="password"
                              value=""
                    />
          </div>
          <div>
                    <input
                              type="submit"
                              value="Register"
                    />
          </div>
</form>,
      ],
    },
    "ref": null,
    "rendered": Array [
      false,
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <div>
              <input
                            onChange={[Function]}
                            placeholder="Username"
                            type="text"
                            value=""
              />
</div>,
            <div>
              <input
                            onChange={[Function]}
                            placeholder="Email"
                            type="email"
                            value=""
              />
</div>,
            <div>
              <input
                            onChange={[Function]}
                            placeholder="Password"
                            type="password"
                            value=""
              />
</div>,
            <div>
              <input
                            onChange={[Function]}
                            placeholder="Confirm Password"
                            type="password"
                            value=""
              />
</div>,
            <div>
              <input
                            type="submit"
                            value="Register"
              />
</div>,
          ],
          "className": "form",
          "onSubmit": [Function],
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": Array [
                <input
                  onChange={[Function]}
                  placeholder="Username"
                  type="text"
                  value=""
/>,
                false,
              ],
            },
            "ref": null,
            "rendered": Array [
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "onChange": [Function],
                  "placeholder": "Username",
                  "type": "text",
                  "value": "",
                },
                "ref": null,
                "rendered": null,
                "type": "input",
              },
              false,
            ],
            "type": "div",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": Array [
                <input
                  onChange={[Function]}
                  placeholder="Email"
                  type="email"
                  value=""
/>,
                false,
              ],
            },
            "ref": null,
            "rendered": Array [
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "onChange": [Function],
                  "placeholder": "Email",
                  "type": "email",
                  "value": "",
                },
                "ref": null,
                "rendered": null,
                "type": "input",
              },
              false,
            ],
            "type": "div",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": Array [
                <input
                  onChange={[Function]}
                  placeholder="Password"
                  type="password"
                  value=""
/>,
                false,
              ],
            },
            "ref": null,
            "rendered": Array [
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "onChange": [Function],
                  "placeholder": "Password",
                  "type": "password",
                  "value": "",
                },
                "ref": null,
                "rendered": null,
                "type": "input",
              },
              false,
            ],
            "type": "div",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": Array [
                <input
                  onChange={[Function]}
                  placeholder="Confirm Password"
                  type="password"
                  value=""
/>,
                false,
              ],
            },
            "ref": null,
            "rendered": Array [
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "onChange": [Function],
                  "placeholder": "Confirm Password",
                  "type": "password",
                  "value": "",
                },
                "ref": null,
                "rendered": null,
                "type": "input",
              },
              false,
            ],
            "type": "div",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": <input
                type="submit"
                value="Register"
/>,
            },
            "ref": null,
            "rendered": Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "type": "submit",
                "value": "Register",
              },
              "ref": null,
              "rendered": null,
              "type": "input",
            },
            "type": "div",
          },
        ],
        "type": "form",
      },
    ],
    "type": "section",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          false,
          <form
            className="form"
            onSubmit={[Function]}
>
            <div>
                        <input
                                    onChange={[Function]}
                                    placeholder="Username"
                                    type="text"
                                    value=""
                        />
            </div>
            <div>
                        <input
                                    onChange={[Function]}
                                    placeholder="Email"
                                    type="email"
                                    value=""
                        />
            </div>
            <div>
                        <input
                                    onChange={[Function]}
                                    placeholder="Password"
                                    type="password"
                                    value=""
                        />
            </div>
            <div>
                        <input
                                    onChange={[Function]}
                                    placeholder="Confirm Password"
                                    type="password"
                                    value=""
                        />
            </div>
            <div>
                        <input
                                    type="submit"
                                    value="Register"
                        />
            </div>
</form>,
        ],
      },
      "ref": null,
      "rendered": Array [
        false,
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <div>
                <input
                                onChange={[Function]}
                                placeholder="Username"
                                type="text"
                                value=""
                />
</div>,
              <div>
                <input
                                onChange={[Function]}
                                placeholder="Email"
                                type="email"
                                value=""
                />
</div>,
              <div>
                <input
                                onChange={[Function]}
                                placeholder="Password"
                                type="password"
                                value=""
                />
</div>,
              <div>
                <input
                                onChange={[Function]}
                                placeholder="Confirm Password"
                                type="password"
                                value=""
                />
</div>,
              <div>
                <input
                                type="submit"
                                value="Register"
                />
</div>,
            ],
            "className": "form",
            "onSubmit": [Function],
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": Array [
                  <input
                    onChange={[Function]}
                    placeholder="Username"
                    type="text"
                    value=""
/>,
                  false,
                ],
              },
              "ref": null,
              "rendered": Array [
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "host",
                  "props": Object {
                    "onChange": [Function],
                    "placeholder": "Username",
                    "type": "text",
                    "value": "",
                  },
                  "ref": null,
                  "rendered": null,
                  "type": "input",
                },
                false,
              ],
              "type": "div",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": Array [
                  <input
                    onChange={[Function]}
                    placeholder="Email"
                    type="email"
                    value=""
/>,
                  false,
                ],
              },
              "ref": null,
              "rendered": Array [
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "host",
                  "props": Object {
                    "onChange": [Function],
                    "placeholder": "Email",
                    "type": "email",
                    "value": "",
                  },
                  "ref": null,
                  "rendered": null,
                  "type": "input",
                },
                false,
              ],
              "type": "div",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": Array [
                  <input
                    onChange={[Function]}
                    placeholder="Password"
                    type="password"
                    value=""
/>,
                  false,
                ],
              },
              "ref": null,
              "rendered": Array [
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "host",
                  "props": Object {
                    "onChange": [Function],
                    "placeholder": "Password",
                    "type": "password",
                    "value": "",
                  },
                  "ref": null,
                  "rendered": null,
                  "type": "input",
                },
                false,
              ],
              "type": "div",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": Array [
                  <input
                    onChange={[Function]}
                    placeholder="Confirm Password"
                    type="password"
                    value=""
/>,
                  false,
                ],
              },
              "ref": null,
              "rendered": Array [
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "host",
                  "props": Object {
                    "onChange": [Function],
                    "placeholder": "Confirm Password",
                    "type": "password",
                    "value": "",
                  },
                  "ref": null,
                  "rendered": null,
                  "type": "input",
                },
                false,
              ],
              "type": "div",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": <input
                  type="submit"
                  value="Register"
/>,
              },
              "ref": null,
              "rendered": Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "type": "submit",
                  "value": "Register",
                },
                "ref": null,
                "rendered": null,
                "type": "input",
              },
              "type": "div",
            },
          ],
          "type": "form",
        },
      ],
      "type": "section",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
    "disableLifecycleMethods": true,
  },
}
`;
